readme.md
# Meter Miser
 
 
## Overview:
Small business owners that operate retail storefronts have a lot to do every day.  Remembering to turn down/up the thermostat often falls to the very bottom of a very long "To Do" list.  Yet, forgetting this task can cost this small business owner hundreds or even thousands of dollars per year depending on the number an size of their retail locations.
What if there were a way to intelligently monitor and alert the small business owner when there are opportunities to adjust the thermostat and generate cost savings?
 
 
**Concept:**
* Automated and periodic polling of wi-fi enabled thermostats in each location
* Persistent storage of collected data for analysis and actionable insights
* Mobile first design to leverage the small business owners most powerful tool -- the smart phone
* Dynamic and interactive charts to quickly identify trends and compare locations
* Ability to take action (adjust/turn off) on the thermostat
 
##Github Link:
[MeterMiser]()
 
##Team Members & Roles:
**Click on each member's name to see their GitHub profile**
All team members are students in the [Digital Crafts](https://digitalcrafts.com) September 2017 cohort. This first project applied agile principles to get a MVP completed in a relatively short timeframe.
 
##The Team
* [Scott Anderson](https://https://github.com/YankeeSoccerNut/) 
**Primary team role:** Product owner and backend developer<br />
**Contributions:**  Original concept, backend, and Google Maps integration<br />
**Key code portions:** All the backend polling, data capture, and API for client GET requests.  Google maps integration with dynamic zoom based on marker boundaries.
 
* [Mikayla Kelhofer](https://github.com/mkelhofer/) 
**Primary team role:** Data Analysis & Visualization<br />
**Contributions:** History Graph with Interactive Timescale <br />
**Key code portions:** Data analysis, d3.js charting, dynamic timeline filtering.
 
* [Michael McFarland](https://github.com/mcfarland422) 
**Primary team role:** Front-End Development and User Interface<br />
**Contributions:** Built custom HTML layout using Bootstrap framework and created custom CSS styling for all aspects of application. Scheduled daily meetings and worked alongside with Scrum team on application from concept to completion. Developed visual and UI aspects of MeterMiser.<br />
**Key code portions:** All of the HTML and CSS.  Overall style and flow.<br />
 
 
##What we used:
**Languages:**
* Node
* JavaScript
* HTML5
* CSS
 
**Frameworks:**
* Express
* Bootstrap
* d3.js
 
**API's**
* Google Maps
* Honeywell "red link"
  * [MobileV2](https://tccna.honeywell.com/ws/MobileV2.asmx)
 
**Other:** 
* 1 Photoshop
* 2 CyberDuck
* 3 MySQL
* 4 Trello
* 5 Google Drive
* 5 SourceTree
 
 
##MVP (Minimum Viable Product):
 
* List locations for user
* Map locations on google map
* Provide historical view
* Provide current, realtime snapshot of thermostat settings
* Identify opportunities for cost savings
 
**Stretch Goals**
* Sign up and authentication
* Push notifications of savings opportunities
* SMS-based interactions
 
## Challenges & Solutions:
**Some of the biggest challenges we faced with this project build included:**
 
<br />1.  **Challenge:** Back-end implementation for polling, persistent storage and client APIs
 
    **Solution:**  We had some team member experiences with relational databases and SQL but not with any particular backend technologies.  After consultation with the DigitalCrafts instructor, we chose Express as the "front-end" to our back-end services.  This choice enabled us to leverage javascript and have a working solution up quickly.  The fact that the API responds with JSON makes it very easy to consume these services from our front-end client.<br />
 
 
<br />2.  **Challenge:**   Front-End:
    We ran into a few issues regarding elements of bootstrap not working correctly. Toggles would not toggle, drop down menus did not function properly, and CDNs were not responding as anticipated.
 
    **Solution:**<br />
 
    These issues were resolved by spending hours upon hours searching the internet and developer communities such as Stackoverflow.com, and talking with our instructor Rob.  we found solutions to all of
    the front-end issues by using these methods and was successfully able to implement them into our application.
 
<br />3.  **Challenge:**  Learning D3.js
  D3.js has a steep learning curve and many resources were with older versions of the module.
 
    **Solution:**<br />
These issues were resolved going through different reference manuals, working with demo graphs, and with Rob.  We were able to compare v3 and v4 to resolve most issues.  We were able to implement an interactive component by adding a callback function to our event-listeners. <br />
 
 
##Code Snippets
 
<!-- Insert code here -->
We created a front-end service to access the API we built on the back-end. 
 
```JavaScript
//App Service for accessing the API
var getLocationsURL = 'http://ec2-18-221-219-61.us-east-2.compute.amazonaws.com/Locations';
$.get(getLocationsURL, function( locations, status ) {....
```
 
## Screenshots:
![Login](https://github.com/YankeeSoccerNut/MeterMiser/blob/master/Screen%20Shots/Screen%20Shot%202017-10-24%20at%2012.18.04%20PM.png)
![Locations](https://github.com/YankeeSoccerNut/MeterMiser/blob/master/Screen%20Shots/Screen%20Shot%202017-10-24%20at%2012.18.59%20PM.png)
![History](https://github.com/YankeeSoccerNut/MeterMiser/blob/master/Screen%20Shots/Screen%20Shot%202017-10-24%20at%2012.19.50%20PM.png)
>
 
 
##Project History
10/24/2017 - Demo MVP 
10/17/2017 - Project Start 


meterMaid/history.txt
    1  exit
    2  ls
    3  cd Desktop
    4  cd DigitalCrafts
    5  ls
    6  cd unit1
    7  ls
    8  cd linux101
    9  cd /
   10  cd ~
   11  ls -al .*
   12  ls .zsh*
   13  cp .zshrc .zshrc.backup
   14  atom .zshrc
   15  man locate
   16  locate .zshrc
   17  locate -i .zshrc
   18  cd Desktop
   19  cd DigitalCrafts
   20  ls
   21  cd unit1
   22  ls
   23  python python101.py
   24  ls
   25  mkdir python101
   26  mv python101.py python101
   27  cd python101
   28  ls
   29  python
   30  python cond.py
   31  python guessTheNumber.py
   32  cp guessTheNumber.py guessTheNumber2.py
   33  python guessTheNumber2.py
   34  cp guessTheNumber2.py guessTheNumber3.py
   35  python guessTheNumber3.py
   36  cp guessTheNumber3.py guessTheNumber4.py
   37  python guessTheNumber4.py
   38  cp guessTheNumber4.py guessTheNumber5.py
   39  python guessTheNumber5.py
   40  ls -al
   41  git init
   42  git remote add origin https://github.com/YankeeSoccerNut/python101.git\ngit push -u origin master
   43  git add .
   44  git remote add origin https://github.com/YankeeSoccerNut/python101.git\ngit push -u origin master
   45  git status
   46  git commit -m "first commit"
   47  git remote add origin https://github.com/YankeeSoccerNut/python101.git\ngit push -u origin master
   48  ls
   49  python guessTheNumber5.py
   50  python dayofweek.py
   51  git status
   52  git add .
   53  git status
   54  exit
   55  pwd
   56  cd Desktop/DigitalCrafts/unit1/python101
   57  ls
   58  python dayofweek.py
   59  python helloyou.py
   60  cd ~
   61  ls
   62  cd Sites
   63  ls
   64  ls -al
   65  cat .localized
   66  cd ..
   67  cat scott.anderson@oracle.com.ams
   68  cd PythonHardway
   69  ls
   70  touch test.txt
   71  rm test.txt
   72  ls
   73  cd Desktop/DigitalCrafts
   74  ls
   75  cd unit1
   76  ls
   77  cd pyt
   78  cd python101
   79  git status
   80  git add .
   81  git status
   82  git commit -m "Added Extra python"
   83  git push
   84  exit
   85  ls
   86  cd Desktop/DigitalCrafts/unit1
   87  ls
   88  cd python101
   89  cd ~/Down
   90  cd ~/Downloads
   91  ls
   92  cd ../Desktop
   93  ls
   94  cd DigitalCrafts/unit1/python101
   95  ls
   96  exit
   97  echo $0
   98  cd ..
   99  ls
  100  exit
  101  ls
  102  cd Desktop/DigitalCrafts/unit1/python101
  103  ls
  104  python guessTheNumber5.py
  105  ls
  106  exit
  107  ls
  108  cd Desktop/DigitalCrafts/unit1/python101
  109  touch testfile
  110  rm testfile
  111  python loops.py
  112  ls
  113  ls -la
  114  cd .git
  115  ls
  116  cat config
  117  cd hooks
  118  ls
  119  cat update.sample
  120  cd ..
  121  ls
  122  python loops.py
  123  git status
  124  python function_practice.py
  125  python dc_challenge.py
  126  python function_practice.py
  127  python dc_challenge.py
  128  pwd
  129  python string.py
  130  python cipher.py
  131  python lists.py
  132  git status
  133  git add .
  134  git commit -m "Lots of exercises"
  135  git push
  136  python
  137  python lists.py
  138  python
  139  python lists.py
  140  pwd
  141  python primeFactor.py
  142  python dictionaries.py
  143  pwd
  144  cd ..
  145  ls
  146  mkdir goblingame
  147  ls
  148  pip install pygame
  149  sudo pip install pygame
  150  sudo -H pip install pygame
  151  python
  152  ls
  153  cd goblingame
  154  ls
  155  python game.py
  156  mkdir images
  157  mv ~/Downloads/images ./images
  158  cd images
  159  ls
  160  pwd
  161  ls
  162  cd images
  163  ls
  164  pwd
  165  mv *.* ..
  166  ls
  167  cd ..
  168  ls
  169  rmdir images
  170  cd images
  171  ls
  172  ls -al
  173  rm *.*
  174  ls
  175  ls -al
  176  cd ..
  177  python game.py
  178  pwd
  179  ls
  180  cd images
  181  ls
  182  rm -r images
  183  ls
  184  cd ..
  185  python game.py
  186  pwd
  187  python game.py
  188  cd Desktop/DigitalCrafts/unit1/python101
  189  ls
  190  python game.py
  191  cd Desktop/DigitalCrafts/unit1
  192  ls
  193  cd goboingame
  194  cd goblingame
  195  ls
  196  python game.py
  197  cd ..
  198  cd python101
  199  python primeFactor.py
  200  sudo python primeFactor.py
  201  python primeFactor.py
  202  cd ..
  203  cd goblingame
  204  python game.py
  205  cd images
  206  ls
  207  cd ..
  208  ls
  209  cd ..
  210  ls
  211  mv LordVader2.png ./unit1/goblingame/images
  212  ls
  213  cd unit1/goblingame
  214  cd images
  215  ls
  216  rm lordvader.png
  217  mv LordVader2.png LordVader.png
  218  cd ..
  219  python game.py
  220  cd Desktop/DigitalCrafts/unit1/goblingame
  221  python game.py
  222  cd images
  223  ls
  224  cd ..
  225  python game.py
  226  cd Desktop/DigitalCrafts/unit1/goblingame
  227  python game.py
  228  exit
  229  cd Desktop/DigitalCrafts/unit1/goblingame
  230  python game.py
  231  cd images
  232  ls
  233  cd ..
  234  ls
  235  python game.py
  236  cd Desktop/DigitalCrafts/unit1/goblingame
  237  python game.py
  238  exit
  239  cd Desktop/DigitalCrafts/unit1/goblingame
  240  python game.py
  241  exit
  242  cd Desktop/DigitalCrafts/unit1/goblingame
  243  python game.py
  244  cd Desktop/DigitalCrafts/unit1/gob
  245  ls
  246  cd gob
  247  pwd
  248  cd Desktop/DigitalCrafts/unit1
  249  ls
  250  cd goblingame
  251  python game.py
  252  cd Desktop/DigitalCrafts/unit1/ls
  253  ls
  254  cd Desktop/DigitalCrafts/unit1
  255  ls
  256  mv goblingame spaceEvader
  257  cd spaceEvader
  258  ls
  259  p
  260  python game.py
  261  git init
  262  git add .
  263  git commit -m "first commit"
  264  git status
  265  git remote add origin https://github.com/YankeeSoccerNut/spaceEvader.git\ngit push -u origin master
  266  python game.py
  267  cd Desktop/DigitalCrafts/unit1
  268  ls
  269  cd spaceEvader
  270  python game.py
  271  cd Desktop/DigitalCrafts/unit1
  272  cd spaceEvader
  273  python game.py
  274  cd Desktop/DigitalCrafts/unit1/spaceEvader
  275  python game.py
  276  git status
  277  git add .

TestD3/d3scott.js
$(document).ready(()=>{

  // hard coding SVG size for now...TODO:  make it dynamic
  // 1. set the dimensions and margins of the graph
  var margin = {top: 30, right: 20, bottom: 70, left: 50},
      width = 1920 - margin.left - margin.right,
      height = 600 - margin.top - margin.bottom;

  // 2. parse the date / time
  var strictIsoParse = d3.utcParse("%Y-%m-%dT%H:%M:%S.%LZ");

  // 3. set the ranges
  var x = d3.scaleTime().range([0, width]);
  var y = d3.scaleLinear().range([height, 0]);

  // 4. Define the axes
  var xAxis = d3.axisBottom(x).tickFormat(function(d){ return d.thermoStatId;});

  var yAxis = d3.axisLeft(y)
      .ticks(10);

  // 5.  define the line
  var myLine = d3.line()
    .x(function(d) { return x(d.thermCreated); })
    .y(function(d) { return y(d.dispTemp); });

  // 6.  append the svg obgect to the body of the page
  // appends a 'group' element to 'svg'
  // moves the 'group' element to the top left margin
  var svg = d3.select("body").append("svg")
      .attr("width", width + margin.left + margin.right)
      .attr("height", height + margin.top + margin.bottom)
    .append("g")
      .attr("transform",
            "translate(" + margin.left + "," + margin.top + ")");

  // 7. Get the readings data

  var getReadingsURL = 'http://ec2-18-221-219-61.us-east-2.compute.amazonaws.com/Readings';

  d3.json(getReadingsURL, function(error, data) {
    // 8.  format the data
      data.forEach(function(d) {
          d.thermCreated = strictIsoParse(d.thermCreated);
          d.dispTemp = +d.dispTemp;
      });

      // 9.  Scale the range of the data
      x.domain(d3.extent(data, function(d) { return d.thermCreated; }));
      y.domain([60, d3.max(data, function(d) { return d.dispTemp; })]);

      // 10.  Nest the entries by thermostatId
      var dataNest = d3.nest()
          .key(function(d) {return d.thermostatId;})
          .entries(data);

      // 11.  Loop through each key (thermostatId)
      // lot going on here....
      var color = d3.scaleOrdinal(d3.schemeCategory10);
      var legendSpace = width/dataNest.length;  // to add legend

      dataNest.forEach(function(d, i) {  // d is data point, i is the index value
          svg.append("path")
              .attr("class", "line")
              .style("stroke", function() {
                return d.color = color(d.key); })
              .attr("d", myLine(d.values));

              // Add the Legend
           svg.append("text")
               .attr("x", (legendSpace/2)+i*legendSpace)
               .attr("y", height + (margin.bottom/2)+ 5)
               .attr("class", "legend")
               .style("fill", function() {
                   return d.color = color(d.key); })
               .text(d.key);
      });  // dataNest.forEach

      // 12.  Add the x Axis
      svg.append("g")
          .attr("class", "x axis")
          .attr("transform", "translate(0," + height + ")")
          .call(xAxis);


      // 13.  Add the y Axis
      svg.append("g")
          .attr("class", "y axis")
          .call(yAxis);

      // text label for the y axis

  });  // d3 getReadings request
}); // document ready
